### PR TITLE
Discard individual selection of lines from file

### PR DESCRIPTION
**Pull request by rafeca**
Closes https://github.com/desktop/desktop/issues/2402

## Description

This PR allows users to discard local changes to specific lines or hunks from a file via the context menu. To do so, a new `git` method called `discardChangesFromSelection()` has been created.

The implementation of the `discardChangesFromSelection()` method differs from the one initially done in https://github.com/desktop/desktop/pull/9399 : instead of reading the whole file into memory and manually copying/deleting the discarded lines, we rely on `git` and its patching functionality to discard the changes. This should make the discard process much more performant for large files.

In order to do so, a new method `formatPatchToDiscardChanges()` has been created. This method generates a string containing the patch that needs to be applied to discard the selected changes from the diff.

### Screenshots

demo

## Release notes

Notes: [Added] Discard individual lines from a changed file

### CODE REVIEW

**Comment by rafeca** [comment]
I'm going to add a bunch of unit tests to the `discardChangesFromSelection()` method, since I had to deal with many edge cases while implementing this.

@tierninho would be good to do quite a bit of testing of discarding lines/hunks from files on different scenarios. Some stuff I tested so far:

* Files marked as moved.
* Files with some changes added to the index.
* Deleted/added files.

There is probably more tricky scenarios (or combinations of these) that would be worth testing.

**Comment by rafeca** [comment]
I've added 4675a23 which does a few changes:

1. Changes the label names to remove the line number from there (sometimes displaying the line number gets confusing since the diff tool display two different sets of line numbers - one for the original file and another for the modified one).
2. Disables the "Discard this line" menu item on hunks that have both added and removed lines. This is done to avoid an unexpected behaviour when discarding changes from modified lines.

As an example of 2., if a user tries to discard the added line L275 from the following chunk:

image

Then the expected behaviour would be to remove the comment from that line. Instead, what `git` will do is removing that added line and end up with this line completely removed in the file:

image

**Comment by rafeca** [comment]
Thanks for testing this @tierninho ! ✨ 

> Right clicking in the "discard all lines" area of a file that has only one individual line says: “Discard these [added/deleted] lines”. If the file has only one line, should we change the language to singular?

Yup, the language should be in singular in this case. I'm trying to reproduce with a new created file with a single line but I'm seeing a correct message:

image

When do you see the wrong message?

> If a file is modified and the user opts to delete all the new lines, the entire contents of the file are deleted. For example if I click "discard all lines" in this file initiated from either the red or green area, then all file contents are deleted, including the white area. I think this ok but just want to point it out.

We shouldn't delete all file contents when discarding changes, that is quite wrong 😟 Are you sure that all the file contents were deleted when you did so (tried reproducing but couldn't)? Maybe it's just that the file had no longer any local change and then Desktop didn't show any changes on the diff view.

> I discarded some lines, edited the file in Atom again, and then tried to discard some more lines but the single-line discard option was greyed out. Why?

We've decided to disable discarding single lines on hunks that contain both added and removed lines, since discarding lines in these scenarios can cause unexpected results (see comment in https://github.com/desktop/desktop/pull/9728#issuecomment-625460839). Was it confusing that you couldn't discard single lines there?

> Do we want to leave the user in a state with an empty file as the user can delete all added/deleted lines?

Good point! Yup I think we should delete the file in that case, gonna implement that.

> If a user removes modified lines and then file disappears, is this expected?

Do you mean disappearing from the GitHub Desktop UI? If that's the case then yes: if a user discards all changed lines of a file, it'll disappear from the "changed files" list since it won't have local changes anymore.

> The blue highlighting was somewhat confusing at first. A line does not have to be selected blue in order to right-click. I understand why, it was just odd looking like this:

I agree, the fact that line numbers can be used both to select changes and discard lines can be a bit confusing at first. 

> "Undo" button would be nice, but I foresee problems if 100 lines get reinserted. The 'Are you sure?' popup is a good compromise.

I think we'll have to end up implementing an "undo" button at some point, but we need to find a good place in the current UI to fit it.

> I managed to discard lines from a conflicted file. I believe we are supporting this.

ups, I didn't test this 😅 did you see any unexpected result when doing so? I'll do some tests and if things weird happen we can disable that.

**Comment by tierninho** [comment]
@rafeca Thanks for the answers! 

> Yup, the language should be in singular in this case. I'm trying to reproduce with a new created file with a single line but I'm seeing a correct message. When do you see the wrong message?

I cannot reproduce it now so let's ignore this for now.

> We shouldn't delete all file contents when discarding changes, that is quite wrong 😟 Are you sure that all the file contents were deleted when you did so (tried reproducing but couldn't)? Maybe it's just that the file had no longer any local change and then Desktop didn't show any changes on the diff view.

Yes, sorry that is what I meant --only the local changes were discarding. The existing file remained as-is, so we are ok here.

> We've decided to disable discarding single lines on hunks that contain both added and removed lines, since discarding lines in these scenarios can cause unexpected results (see comment in #9728 (comment)). Was it confusing that you couldn't discard single lines there?

Only somewhat confusing at first but I realized that I had made multiple rounds of changes and that must of been the cause. 

> Do you mean disappearing from the GitHub Desktop UI? If that's the case then yes: if a user discards all changed lines of a file, it'll disappear from the "changed files" list since it won't have local changes anymore.

Yes, from the UI. Thanks for the confirmation.

> ups, I didn't test this 😅 did you see any unexpected result when doing so? I'll do some tests and if things weird happen we can disable that.

All seemed to work fine when I tested it 👍 I'll report back if I can break it.

**Comment by tierninho** [comment]
I was just testing merge conflicts again and this conversation👇 made me wonder:

> Do we want to leave the user in a state with an empty file as the user can delete all added/deleted lines?

> Good point! Yup I think we should delete the file in that case, gonna implement that.

Based on that, if this gets implemented should be still be deleting the file even if it is a conflicted file? I think that might cause some problems as a user might remove all lines to resolve a merge conflict and still expect to see the file checked green, like so:

Screen Shot 2020-05-11 at 9 40 31 AM

**Comment by rafeca** [comment]
> Based on that, if this gets implemented should be still be deleting the file even if it is a conflicted file?

Good point, in the case of a conflicted file, it should still appear on the changed files list, even if the user discard all its changes.

File: app/src/lib/feature-flag.ts [hunk 1]
```diff
@@ -185,3 +185,10 @@ export function enableGitTagsDisplay(): boolean {
 export function enableGitTagsCreation(): boolean {
   return true
 }
+
+/**
+ * Should we show the discard lines/hunks contet menu item?
```
Comment (Rexogamer) [inline app/src/lib/feature-flag.ts:190]: tiny typo:
```suggestion
 * Should we show the discard lines/hunks context menu item?
```

File: app/src/ui/diff/text-diff.tsx [hunk 2]
```diff
@@ -480,17 +499,138 @@ export class TextDiff extends React.Component<ITextDiffProps, {}> {
...
+
+    const selection = file.selection
+      .withSelectNone()
+      .withRangeSelection(startLine, endLine - startLine + 1, true)
+
+    this.props.onDiscardChanges(this.props.diff, selection)
+  }
+
+  private getDiscardLabel(hunkType: HunkType, numLines: number): string {
+    return toPlatformCase(
```
Comment (niik) [inline app/src/ui/diff/text-diff.tsx:627]: I'd rather we not use `toPlatformCase`. It has a checkered past in Desktop and I believe the last time we discussed it we were all in favor of removing it so adding more usage to it would be counterproductive.

For context the rules for title casing are much too complex to succinctly express in a helper function like this which is why you see these ternary statements scattered about the code base. See https://github.com/desktop/desktop/pull/5841#pullrequestreview-162478757 and https://github.com/desktop/desktop/issues/5847 for some of the back story.

In this scenario it's possible that it's doing the right thing but while I'm far far far from an expert in English grammar I wonder if the word "these" can be considered an article in this sentence and therefore should be excluded from uppercasing based on https://developer.apple.com/library/archive/documentation/FinalCutProX/Conceptual/FxPlugHIG/TextStyleGuidelines/TextStyleGuidelines.html

> In title-style capitalization, capitalize every word except:
> - Articles (a, an, the)
>

I propose we make this easy on us by rewording and refactor here 👇 

```ts
private getDiscardLabel(hunkType: HunkType, numLines: number): string {
  const s = numLines > 1 ? 's' : ''
  return __DARWIN__ ? `Discard Line${s}` : `Discard line${s}`
}
```

or, slightly more verbose

```ts
private getDiscardLabel(hunkType: HunkType, numLines: number): string {
  if (numLines > 1) {
    return __DARWIN__ ? 'Discard Lines' : 'Discard lines'
  } else {
    return __DARWIN__ ? 'Discard Line' : 'Discard line'
  }
}
```

File: app/src/ui/diff/text-diff.tsx [hunk 3]
```diff
@@ -773,7 +913,12 @@ export class TextDiff extends React.Component<ITextDiffProps, {}> {
       return
     }
 
-    const { file, hunks, readOnly } = this.props
+    // Discard right click events
```
Comment (niik) [inline app/src/ui/diff/text-diff.tsx:916]: This comment read to me as  "ignore right click events". Might I suggest the following rewording?

```suggestion
    // We only care about the primary button here, secondary
    // button clicks are handled by `onContextMenu`
```

File: app/src/ui/diff/text-diff.tsx [hunk 4]
```diff
@@ -401,6 +415,11 @@ export class TextDiff extends React.Component<ITextDiffProps, {}> {
   private onDocumentMouseUp = (ev: MouseEvent) => {
     ev.preventDefault()
 
+    // Discard right click events
```
Comment (niik) [inline app/src/ui/diff/text-diff.tsx:418]: This comment read to me as  "ignore right click events". Might I suggest the following rewording?

```suggestion
    // We only care about the primary button here, secondary
    // button clicks are handled by `onContextMenu`
```

File: app/src/ui/diff/text-diff.tsx [hunk 5]
```diff
@@ -797,7 +942,12 @@ export class TextDiff extends React.Component<ITextDiffProps, {}> {
       return
     }
 
-    const { file, hunks, readOnly } = this.props
+    // Discard right click events
```
Comment (niik) [inline app/src/ui/diff/text-diff.tsx:945]: This comment read to me as  "ignore right click events". Might I suggest the following rewording?

```suggestion
    // We only care about the primary button here, secondary
    // button clicks are handled by `onContextMenu`
```

File: app/src/lib/patch-formatter.ts [hunk 6]
```diff
@@ -225,3 +225,106 @@ export function formatPatch(
...
+      }
+
+      // Context lines can always be let through, they will
+      // never appear for new files.
+      if (line.type === DiffLineType.Context) {
+        hunkBuf += `${line.text}\n`
+        oldCount++
+        newCount++
+      } else if (selection.isSelected(absoluteIndex)) {
+        // Reverse the change (if it was an added line, treat is as removed and viceversa).
```
Comment (niik) [inline app/src/lib/patch-formatter.ts:276]: total nitpick

```suggestion
        // Reverse the change (if it was an added line, treat it as removed and vice versa).
```

File: app/src/lib/stores/git-store.ts [hunk 7]
```diff
@@ -1497,6 +1499,21 @@ export class GitStore extends BaseStore {
     })
   }
 
+  public async discardChangesFromSelection(
+    filePath: string,
+    diff: ITextDiff,
+    selection: DiffSelection
+  ) {
+    await this.performFailableOperation(async () => {
+      await discardChangesFromSelection(
```
Comment (niik) [inline app/src/lib/stores/git-store.ts:1508]: I think we can skip the inner await here since `performFailableOperation` will await it so it only creates another Promise closure 

```suggestion
      discardChangesFromSelection(
```

File: app/src/ui/diff/text-diff.tsx [hunk 8]
```diff
@@ -480,17 +499,138 @@ export class TextDiff extends React.Component<ITextDiffProps, {}> {
...
+          action: () => this.onDiscardChanges(file, lineNumber),
+          enabled: hunkType !== HunkType.Modified,
+        },
+      ]
+    }
+
+    return null
+  }
+
+  private getHunkType(range: { from: number; to: number }): HunkType | null {
```
[the remark this answers is not in the document]
Comment (rafeca) [reply app/src/ui/diff/text-diff.tsx:589]: Makes sense! I think the naming that you proposed is much better than the current one as well 😅 

re. checking each line, I tried to avoid having to do so because new added files (or removed ones) there's only going to be a single hunk (or range composition) and we'll need to iterate through all file lines. Anyways since the check is super simple and it's only executed when right clicking I don't think it should have any noticeable impact on performance (I'm going to do a quick check with a file with a few thousand lines just to be sure).

File: app/src/ui/diff/text-diff.tsx (hunk 2 above)
Comment (rafeca) [reply app/src/ui/diff/text-diff.tsx:627]: Good point! I just realized that this method was not used anywhere else in the app 🤦‍♂️ (don't know how I even discovered it).

If you don't mind I'll send another PR to completely remove that method from the codebase.

File: app/src/lib/stores/git-store.ts (hunk 7 above)
Comment (rafeca) [reply app/src/lib/stores/git-store.ts:1508]: I don't think we can just remove the `await` here: we'll need to also return the promise that gets returned by `discardChangesFromSelection()` to make `performFailableOperation()` await that promise.

I can do that by removing the braces from the function and get advantage of the implicit return.

File: app/src/lib/git/discard-changes.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+ * @param repository The repository in which to update the working directory
+ *                   with information from the index
+ *
+ * @param filePath   The relative path in the working directory of the file to use
+ *
+ * @param diff       The diff containing the file local changes
+ *
+ * @param selection  The selection of changes from the diff to discard
+ */
+export async function discardChangesFromSelection(
```
Comment (rafeca) [inline app/src/lib/git/discard-changes.ts:22]: @niik I'm going to move this method to the `apply.ts` file, since as we discussed we use the first git argument as the name of the file for git methods.

File: app/src/ui/diff/text-diff.tsx (hunk 2 above)
Comment (niik) [reply app/src/ui/diff/text-diff.tsx:628]: > If you don't mind I'll send another PR to completely remove that method from the codebase.

I don't mind, quite the opposite 😄

File: app/src/lib/git/discard-changes.ts (hunk 9 above)
Comment (niik) [reply app/src/lib/git/discard-changes.ts:22]: Awesome!

File: app/src/ui/diff/text-diff.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (rafeca) [reply app/src/ui/diff/text-diff.tsx:590]: I've done some tests with out current `yarn.lock` file which is 11k lines (I think this could be a relatively common scenario, where a user wants to discard a `yarn.lock` file), and while the results are worse than I expected they're still not terribly bad.

It took between 2ms in the best case (probably with all codecaches warned in v8) and 20ms in the worse case (these results are on an average Intel i5 Windows laptop).

Something a bit worrying is that we have to call `diffLineForIndex()` for each line in the range, and this method iterates through all hunks on the file, which can potentially lead to an `O(n * m)` operation (it's a quite rare scenario, since this would require selecting a hunk with a ton of lines inside a file with a ton of hunks, but still something to be aware).

Main problem here is `getHunkType()` accepts a range, which can potentially contain lines from multiple hunks, although we know for sure (if I'm not wrong) that in the discard scenario you can only discard a hunk at a time. I'm going to try to change slightly the API of this method to be more restrictive and avoid this `O(n * m)` operation

[the remark this answers is not in the document]
Comment (rafeca) [reply app/src/ui/diff/text-diff.tsx]: I've ended up addressing this in 754f85b49: by moving this calculation to the `findInteractiveDiffRange()` method we make it costless (at least in big o terms).

This could add (an unapreciable tbh) overhead to other methods using `findInteractiveDiffRange()`, but since this is only used for calculating selections on partial commits it should be fine.

**Comment by rafeca** [comment]
@niik I've addressed your comments so this PR is ready for another review 😃

File: app/src/ui/diff/diff-explorer.ts [hunk 10]
```diff
@@ -1,4 +1,27 @@
-import { DiffLine, DiffHunk } from '../../models/diff'
+import { DiffLine, DiffHunk, DiffLineType } from '../../models/diff'
+
+/**
+ * Indicate the type of changes that are included in the current range.
+ */
+export enum DiffRangeType {
+  /** Only contains added lines. */
+  Additions,
+  /** Only contains added lines. */
```
Comment (niik) [inline app/src/ui/diff/diff-explorer.ts:9]:
```suggestion
  /** Only contains deleted lines. */
```

File: app/src/ui/diff/index.tsx [hunk 11]
```diff
@@ -64,6 +64,11 @@ interface IDiffProps {
 
   /** Hiding whitespace in diff. */
   readonly hideWhitespaceInDiff: boolean
+
+  readonly onDiscardChanges?: (
```
Comment (niik) [inline app/src/ui/diff/index.tsx:68]: Can we copy and paste the nice documentation you added for `onDiscardChanges` in `ITextDiffProps` here?

File: app/src/ui/diff/diff-explorer.ts [hunk 12]
```diff
@@ -85,10 +115,52 @@ export function findInteractiveDiffRange(
...
   }
 
   const to =
     contextLineAfterIndex !== null ? contextLineAfterIndex : hunk.unifiedDiffEnd
 
-  return { from, to }
+  return { from, to, type: rangeType }
+}
+
+function getNextRangeType(
```
Comment (niik) [inline app/src/ui/diff/diff-explorer.ts:128]: Big fan of this approach ✨

File: app/src/ui/diff/diff-explorer.ts [hunk 13]
```diff
@@ -1,4 +1,27 @@
...
+  /** Contains both added and removed lines. */
+  Mixed,
+}
+
+/**
+ * Helper object that represents a range of lines in a diff.
+ * Its type represents the type of interactive (added or deleted)
+ * lines that it contains, being null when it has no interactive lines.
+ */
+type DiffRange = {
```
Comment (niik) [inline app/src/ui/diff/diff-explorer.ts:20]: Love that you took the time to break this out into its own type. I think conventionally we'd use an interface here instead of a type but I'm absolutely fine with this, it's such an improvement over the anonymous tuple type.

**Comment by nk250319** [comment]
Of note, the verbiage SourceTree uses is plural for everything: "Discard Selected Lines" in addition to "Discard Hunk" (if the lines are all grouped together).

File: app/src/ui/diff/text-diff.tsx [hunk 14]
```diff
@@ -480,17 +505,133 @@ export class TextDiff extends React.Component<ITextDiffProps, {}> {
...
+    let type = ''
+
+    if (rangeType === DiffRangeType.Additions) {
+      type = __DARWIN__ ? 'Added' : 'added'
+    } else if (rangeType === DiffRangeType.Deletions) {
+      type = __DARWIN__ ? 'Removed' : 'removed'
+    } else if (rangeType === DiffRangeType.Mixed) {
+      type = __DARWIN__ ? 'Modified' : 'modiifed'
+    } else {
+      assertNever(rangeType, `Invlaid range type: ${rangeType}`)
```
Comment (niik) [inline app/src/ui/diff/text-diff.tsx:621]:
```suggestion
      assertNever(rangeType, `Invalid range type: ${rangeType}`)
```

File: app/src/ui/diff/text-diff.tsx [hunk 15]
```diff
@@ -480,17 +505,133 @@ export class TextDiff extends React.Component<ITextDiffProps, {}> {
...
+
+  private getDiscardLabel(rangeType: DiffRangeType, numLines: number): string {
+    let type = ''
+
+    if (rangeType === DiffRangeType.Additions) {
+      type = __DARWIN__ ? 'Added' : 'added'
+    } else if (rangeType === DiffRangeType.Deletions) {
+      type = __DARWIN__ ? 'Removed' : 'removed'
+    } else if (rangeType === DiffRangeType.Mixed) {
+      type = __DARWIN__ ? 'Modified' : 'modiifed'
```
Comment (niik) [inline app/src/ui/diff/text-diff.tsx:619]:
```suggestion
      type = __DARWIN__ ? 'Modified' : 'modified'
```

**Comment by rafeca** [comment]
Thanks for the feedback @nk250319 and @niik ! I agree that "this"/"these" don't add much information in the label (although I'm not a native speaker so take my opinions with a grain of salt).

Let me summarize a couple of constrains/edge cases that we faced when designing this label, so we can take them into account if we want to iterate again through the text of the labels:

* Since discarding changes does the opposite as what the diff is showing (it will delete green/added lines and it will re-add red/removed lines), this can be a bit confusing to a user. Imagine the following example, where discarding these lines will re-add them to the file (this may be obvious for people :
  image
  This made us add the explicit "Added"/"Deleted" adjective on the label ([link to internal slack conversation](https://github.slack.com/archives/C17LP0XU3/p1588795624418900)).

* Since we have two sets of line numbers on a diff (the ones from the previous state and the ones from the next state), in some situations displaying the line numbers (or even the number of lines to discard) can be quite confusing for users. For example in this scenario: 
  image
  This made us remove the line numbers from the labels ([link to slack conversation](https://github.slack.com/archives/C17LP0XU3/p1588870600433400)).

I think we never discussed whether to use "this"/"these", it just came as some glue between the verb "Discard" and "Added Lines"/"Removed Lines".

I think that we can probably just ditch the "this" and just display "Discard Added Line"/"Discard Added Lines" (I don't have a strong opinion about which one I prefer), what do you think @billygriffin / @ampinsk ?

**Comment by ampinsk** [comment]
> I think that we can probably just ditch the "this" and just display "Discard Added Line"/"Discard Added Lines" (I don't have a strong opinion about which one I prefer), what do you think @billygriffin / @ampinsk ?

I don't mind ditching the "this!" 👍

**Comment by tierninho** [comment]
- Ran through another rounds of testing given the recent changes, and things are functioning as expected.

- I wanted to point out the distinction between discarding one line and multiple lines hard to recognize, as we are literally talking about 1 or 2 pixels apart and one letter away from each other (see pics and notice distance of the context menu). I suggest making it clearer, especially due to the destructive nature of the feature. 

  Screen Shot 2020-06-02 at 4 03 12 PM
  Screen Shot 2020-06-02 at 4 02 27 PM

  When discarding or ignoring files, we include the word "all", perhaps `Discard All Added Lines` is more clear.

**Comment by rafeca** [comment]
Thanks for the feedback @tierninho ! I agree the difference it's too subtle, but I think "Discard All Added Lines" can be misleading since it can mean that it will discard all added lines from that file.

What about "Discard Multiple Added Lines"? it feels super verbose so any other alternative will be appreciated!

**Comment by billygriffin** [comment]
I think I'm fine with where we're at after the removal of "this." There don't seem to be any perfect solutions, but where we're at is pretty close. Unless @ampinsk feels strongly about changing the language from here, I think we can get feedback and update verbiage if there's an opportunity to make it more clear once people are using it.

**Comment by jabamaus** [comment]
Hi,

Thanks for the feature - very useful. Not fully working for me though. 'Discard added line' works as expected but 'Discard modified line' is always greyed out in the context menu. What am I missing?

Cheers.

**Comment by Rexogamer** [comment]
> Hi,
> 
> Thanks for the feature - very useful. Not fully working for me though. 'Discard added line' works as expected but 'Discard modified line' is always greyed out in the context menu. What am I missing?
> 
> Cheers.

Hey! I think you might want to [create an issue for that](https://github.com/desktop/desktop/issues/new/choose) with more info (eg your OS, Github Desktop version and more info about your files/diffs).
